src/oic/utils/session_backend.py

```diff
@@ -0,0 +1,293 @@
+
+    This should really not be used in production.
+    """
+
+    def __init__(self):
+        """Create the storage."""
+        self.storage = {}  # type: Dict[str, Dict[str, str]]
+
+    def __setitem__(self, key: str, value: Dict[str, str]) -> None:
+        """Store the session info in the storage."""
+        self.storage[key] = value
+
+    def __getitem__(self, key: str) -> Dict[str, str]:
+        """Retrieve session information based on session id."""
+        return self.storage[key]
+
+    def __delitem__(self, key: str) -> None:
+        """Delete the session info."""
+        del self.storage[key]
+
+    def __contains__(self, key: str) -> bool:
+        return key in self.storage
+
+    def get_by_sub(self, sub: str) -> List[str]:
+        """Return session ids based on sub."""
+        return [
+            sid for sid, session in self.storage.items() if session.get("sub") == sub
+        ]
+
+    def get_by_uid(self, uid: str) -> List[str]:
+        """Return session ids based on uid."""
+        return [
+            sid
+            for sid, session in self.storage.items()
+            if AuthnEvent.from_json(session["authn_event"]).uid == uid
+        ]
+
+    def get(self, attr: str, val: str) -> List[str]:
+        """Return session ids based on attribute name and value."""
+        return [
+            sid for sid, session in self.storage.items() if session.get(attr) == val
+        ]
+
+    def get_uid_by_sub(self, sub):
+        """Return User ids based on sub. Should only be one."""
+        for sid, session in self.storage.items():
+            if session.get("sub") == sub:
+                return AuthnEvent.from_json(session["authn_event"]).uid
+        return None
+
+    def get_uid_by_sid(self, sid):
+        return self.get_uid_by_sub(self.storage[sid]["sub"])
+
+    def get_by_sub_and_(self, sub, key, val):
+        """
+        Given a subject identifier and a key/value pair return the Id of a session
+        that matches those values.
+
+        :param sub: Subjecty Identifier
+        :param key: attribute name
+        :param val: attribute value
+        :return: Session identifier
+        """
+        for sid, session in self.storage.items():
+            if session.get("sub") == sub and session.get(key) == val:
+                return sid
+
+    def update(self, key, attribute, value):
+        """
+        Updates information stored. If the key is not know a new entry will be
+        constructed.
+
+        :param key: Key to the database
+        :param attribute: Attribute name
+        :param value: Attribute value
+        """
+        if key not in self.storage:
+            self.storage[key] = {attribute: value}
+        else:
+            item = self.storage[key]
+            item[attribute] = value
+            self.storage[key] = item
+
+
+def session_update(db, key, attr, val):
+    if isinstance(db, SessionBackend):
+        db.update(key, attr, val)
+    elif isinstance(db, dict):
```

Comment:
The `else` statements should probably also raise `DeprecationWarning` so we can remove them later.

Reply:
OK